### PR TITLE
NetworkBridge: validator (authorities) discovery api

### PR DESCRIPTION
**Pull request by ordian**
Part of #1461 (only current session for now).

My suggestion is to limit `NetworkBridge` to `AuthorityId`s only and to not track the `ValidatorId -> AuthorityId` mapping (already implemented as runtime API call `validator_discovery`), and ask a caller to do that.
This logic could be extracted somewhere to be reused by different subsystems. This separates concerns better IMHO especially when dealing with multiple sessions.

Unresolved:
- [x] Would it be better to return a `Stream` instead of a `Vec` so that we can wait for the first connected validator?
Probably fine as is: https://github.com/paritytech/polkadot/pull/1699#discussion_r486397245
- [x] Alternatively, pass a timeout and return whatever we've got not to wait for all validators to be connected?
https://github.com/paritytech/polkadot/pull/1699#issuecomment-691019240
- [x] What to do if a peer we are waiting for was connected but then disconnected. Should we add it back to pending list or just return a list w/o this peer? Currently the latter is implemented.

Todo:
- [x] Update collation protocol to use `ConnectToAuthorities`
- [x] Handle multiple in-flight requests
- [x] API for revoking discovery requests
- [x] Tests

### CODE REVIEW

File: node/network/bridge/src/lib.rs
Comment (rphmeier) [inline node/network/bridge/src/lib.rs:275]: It should be called `ConnectToValidators`. `Authority` is a term used lower down in the stack

File: node/network/bridge/src/lib.rs
Comment (rphmeier) [inline node/network/bridge/src/lib.rs:200]: No, not necessary

File: node/network/bridge/src/lib.rs
Comment (rphmeier) [inline node/network/bridge/src/lib.rs:134]: The whole priority groups system makes very little sense to me but we could add functions for tweaking it as needed.

File: node/network/bridge/src/lib.rs
Comment (coriolinus) [inline node/network/bridge/src/lib.rs:621]:
```suggestion
				for authority in authorities.iter().cloned() {
```

A larger number of smaller allocations, but my understanding is that the compiler is smart enough to reuse the memory from the previous iteration of the loop, which amortizes the cost and reduces the total allocation required.

File: node/network/bridge/src/lib.rs
Comment (rphmeier) [reply node/network/bridge/src/lib.rs:615]: I do worry about leakage, though, in case some logic isn't triggered in a particular subsystem to do the clean-up.

File: node/subsystem/src/messages.rs
Comment (coriolinus) [inline node/subsystem/src/messages.rs:205]: IMO a Vec is probably fine in this case.

File: runtime/parachains/src/runtime_api_impl/v1.rs
Comment (rphmeier) [inline runtime/parachains/src/runtime_api_impl/v1.rs:281]: I'm adding a `SessionInfo` module that will alleviate this in #1691

File: node/core/runtime-api/src/lib.rs
Comment (drahnr) [inline node/core/runtime-api/src/lib.rs:264]:
```suggestion
				vec![None; idx.len()]
```

File: node/network/bridge/src/lib.rs
Comment (drahnr) [inline node/network/bridge/src/lib.rs:621]: Do we want to limit them to a fixed value anyways? `addresses.into_iter().take(N)`

File: node/network/bridge/src/lib.rs
Comment (drahnr) [inline node/network/bridge/src/lib.rs:734]: This whole code chunk is duplicated above, maybe a candidate for a closure of `fn` :)

File: node/network/bridge/src/lib.rs
Comment (drahnr) [inline node/network/bridge/src/lib.rs:827]: Clever usage of retain :+1:

**Comment by ordian** [comment]
I implemented multiple in-flight requests in b80e050, but handling leakage is tricky. If we have another message for revoking, since validator sets of different requests may overlap (or is it not true?), what happens is one subsystem revokes an id another is waiting on?

Another unresolved question is what to do if a peer are waiting for was connected but then disconnected. Should we add it back to pending list or just return a list w/o this peer?

File: node/network/bridge/src/lib.rs
Comment (ordian) [reply node/network/bridge/src/lib.rs:734]: the existence of duplicated code chunk is dependent on the second unresolved question: https://github.com/paritytech/polkadot/pull/1699#issuecomment-690966735, so I didn't want to introduce a premature abstraction :)

File: node/network/bridge/src/lib.rs
Comment (tomaka) [reply node/network/bridge/src/lib.rs:200]: What I had in mind is instead to do `impl Network for (Arc<NetworkService>, AuthorityDiscoveryService)`.
It doesn't really make sense to be to be generic over the `NetworkService` but not the `AuthorityDiscoveryService`. It's either both or none.

File: node/network/bridge/src/lib.rs
Comment (tomaka) [reply node/network/bridge/src/lib.rs:275]: Also I'd suggest something like `SetConnectedValidators`. It's a state change that we perform, not a one-shot action.

File: node/network/bridge/src/lib.rs
Comment (tomaka) [inline node/network/bridge/src/lib.rs:623]: I don't understand this loop. The addresses returned by `get_addresses_by_authority_id` are also deduplicated.

File: node/network/bridge/src/lib.rs
Comment (ordian) [reply node/network/bridge/src/lib.rs:275]: ATM we call `set_priority_group`, but what we actually want is `add_to_priority_group`, so it's an incremental change, see https://github.com/paritytech/polkadot/pull/1699#discussion_r486389522.

There is an issue with an unbounded growth of priority_group though, it's not clear to how it should handled https://github.com/paritytech/polkadot/pull/1699#issuecomment-690966735.

File: node/network/bridge/src/lib.rs
Comment (ordian) [reply node/network/bridge/src/lib.rs:623]: The loop is not about deduplication, it just collects all multiaddresses. The reason it uses a `HashMap` and not a `Vec` is because `set_priority_group` accepts a `HashSet`. But we don't want to call `set_priority_group` https://github.com/paritytech/polkadot/pull/1699#discussion_r486957653, so it'll be changed.

File: node/network/bridge/src/lib.rs
Comment (ordian) [reply node/network/bridge/src/lib.rs:200]: Yeah, I agree, it doesn't make sense to abstract over `NetworkService`, but not `AuthorityDiscoveryService`.

**Comment by tomaka** [comment]
Note that there's no guarantee that you will actually manage to connect to all validators. In all likelihood you will actually not.

Either you need to add some timeout system, or, a solution which I think is better, is to change the logic of `Action::ConnectToAuthorities`/`ConnectToValidators`.

Its "send_back" parameter should immediately return the list of `PeerId`s of validators that we are already connected to, and then `PeerConnected` should indicate the validator identity of the newly-connected peer.

**Comment by ordian** [comment]
> Note that there's no guarantee that you will actually manage to connect to all validators. In all likelihood you will actually not.
> 
> Either you need to add some timeout system, or, a solution which I think is better, is to change the logic of `Action::ConnectToAuthorities`/`ConnectToValidators`.
> 
> Its "send_back" parameter should immediately return the list of `PeerId`s of validators that we are already connected to, and then `PeerConnected` should indicate the validator identity of the newly-connected peer.

Yeah, see the first two questions in unresolved sestion: https://github.com/paritytech/polkadot/pull/1699#issue-483912440.
An example of usage is https://github.com/paritytech/polkadot/blob/b32dcc44919b995cdf0ea2a18e33200890ce0c78/node/network/collator-protocol/src/collator_side.rs#L240-L262

Knowing `PeerId` is not enough though, we need to wait for them to be connected in order to send messages, unless there is a way to send a message to a not-connected `PeerId` (which I assume is not yet implemented).

**Comment by tomaka** [comment]
> Knowing PeerId is not enough though, we need to wait for them to be connected in order to send messages

I don't think it's a good idea to block the entire collation logic if a single validator is offline, or behind a NAT, or taking a long time to respond, is it?

> unless there is a way to send a message to a not-connected PeerId (which I assume is not yet implemented).

And never going to be implemented. We don't want to infinitely buffer up messages destined to peers that might never be reachable.

**Comment by ordian** [comment]
> > Knowing PeerId is not enough though, we need to wait for them to be connected in order to send messages
> 
> I don't think it's a good idea to block the entire collation logic if a single validator is offline, or behind a NAT, or taking a long time to respond, is it?

sorry, I misread the *we are already connected to* part and thought you suggested return the list of all `PeerId`s.
Agree this is the way to go and is better than timeouts.
 
> > unless there is a way to send a message to a not-connected PeerId (which I assume is not yet implemented).
> 
> And never going to be implemented. We don't want to infinitely buffer up messages destined to peers that might never be reachable.

was referring to this comment https://github.com/paritytech/polkadot/pull/1699#discussion_r486389522
> A specific validator from a specific session index. e.g. for requesting some particular data from them. @tomaka and I have talked about this, and we might sidestep this just by allowing request/responses to reach un-connected peers.

**Comment by ordian** [comment]
> but handling leakage is tricky. If we have another message for revoking, since validator sets of different requests may overlap (or is it not true?), what happens is one subsystem revokes an id another is waiting on?

Unless you have better ideas, I'm going to implement a ref counting as a way to sidestep that, i.e. there will be global `pending: HashMap<AuthorityId, usize>` map, where the value is the number of times this id was requested by subsystems. There will be another message from subsystems to decrement this number (revoking previous discovery requests). 

It won't solve this problem though:
>  in case some logic isn't triggered in a particular subsystem to do the clean-up.

An alternative would be something like https://github.com/stusmall/ttl_cache, but it's not clear if there is a single timeout that would suite all the subsystems.

**Comment by ordian** [comment]
The PR can be reviewed, although it's currently lacks tests.

File: node/network/collator-protocol/src/collator_side.rs
Comment (montekki) [inline node/network/collator-protocol/src/collator_side.rs:288]: I think it would be handier to `clone()` it beforehand?

File: node/network/collator-protocol/src/collator_side.rs
Comment (montekki) [inline node/network/collator-protocol/src/collator_side.rs:293]: Yes, I don't think we should stall here if we cannot connect to anyone

File: node/network/collator-protocol/src/collator_side.rs
Comment (montekki) [inline node/network/collator-protocol/src/collator_side.rs:258]: Again, I think that there sould just be an api to disconnect from nodes we are no longer interested in, not the whole group we connected to.

This would be useful not only here but in the case validators considering to drop evil collators.

File: node/subsystem/src/messages.rs
Comment (ordian) [reply node/subsystem/src/messages.rs:214]: If we revoke a request, it doesn't necessarily mean that we disconnect from all its peers, it just will remove them from the priority group, they can be added back in a new request if the groups intersect. But I see now that the collator implementation operates on diffs, so yeah, we would need to either change either the collator requests or the revoking api.

Comment (ordian) [reply node/subsystem/src/messages.rs:214]: Actually, looking again at the code, the collator requests a connection for the full group, not just the diff, so it's fine as is.

File: node/network/collator-protocol/src/collator_side.rs
Comment (ordian) [reply node/network/collator-protocol/src/collator_side.rs]: That would complicate the caller's side, because currently, the networking request operates on `AuthorityDiscoveryId`, not `ValidatorId`. The mapping is abstracted away in the `connect_to_validators_impl` function, which is ideally to be extracted somewhere to be reused by other subsystems. 

If we introduce another Message `DisconnectFromValidators`, this mapping would have to be maintained by every subsystem.

> This would be useful not only here but in the case validators considering to drop evil collators.

we could just revoke the previous connection request and not request it anymore. We also have `ReportPeer` message for this.

File: node/network/collator-protocol/src/collator_side.rs
Comment (ordian) [reply node/network/collator-protocol/src/collator_side.rs:303]: Hmm, I'm confused now given that the collator handles `PeerConnected` and `PeerDisconnected` events.
`ConnectToValidators` returns the list when the validators are already connected. Is this not what we want?

Otherwie, there is a problem with a race condition if we receive `PeerConnected` message from the `NetworkBridge` before `ConnectToValidators` response.

File: node/subsystem-util/src/validator_discovery.rs
Comment (montekki) [inline node/subsystem-util/src/validator_discovery.rs:134]:
```suggestion
		if self.validator_map.is_empty() {
			return Poll::Ready(None);
		}
		match Pin::new(&mut self.connections).poll_next(cx) {
			Poll::Ready(Some((id, peer_id))) => {
				if let Some(validator_id) = self.validator_map.remove(&id) {
```

File: node/subsystem-util/src/validator_discovery.rs
Comment (montekki) [inline node/subsystem-util/src/validator_discovery.rs:27]:
```suggestion
```

File: node/network/bridge/src/validator_discovery.rs
Comment (ordian) [reply node/network/bridge/src/validator_discovery.rs:252]: I guess target doesn't matter for warning as they are always enabled, but sure
```suggestion
			log::warn!(target: super::TARGET, "AuthorityDiscoveryService returned an invalid multiaddress: {}", e);
```

File: node/network/collator-protocol/src/collator_side.rs
Comment (ordian) [reply node/network/collator-protocol/src/collator_side.rs:588]:
```suggestion
		while let Poll::Ready(msg) = futures::poll!(ctx.recv()) {
```

File: node/subsystem-util/src/validator_discovery.rs
Comment (ordian) [reply node/subsystem-util/src/validator_discovery.rs]: Doesn't seem to work
https://gitlab.parity.io/parity/polkadot/-/jobs/659694#L586

File: node/subsystem/src/messages.rs
Comment (ordian) [inline node/subsystem/src/messages.rs:204]:
```suggestion
```

File: node/network/bridge/src/lib.rs
Comment (rphmeier) [inline node/network/bridge/src/lib.rs:247]: 👍

File: node/network/bridge/src/validator_discovery.rs
Comment (rphmeier) [inline node/network/bridge/src/validator_discovery.rs:37]: This TODO needs resolving or issue filed before merge

File: node/network/bridge/src/validator_discovery.rs
Comment (rphmeier) [inline node/network/bridge/src/validator_discovery.rs:31]: Should each protocol (collation, validation) have its own priority group?

File: node/network/bridge/src/validator_discovery.rs
Comment (ordian) [reply node/network/bridge/src/validator_discovery.rs:37]: I think it's not exposed in the substrate network API yet, can add in a follow-up

File: node/network/bridge/src/validator_discovery.rs
Comment (ordian) [reply node/network/bridge/src/validator_discovery.rs:31]: that's a good question, I'm not sure about this.
From the NetworkBridge perspective there is no difference. Or rather, what would be the benefit of separating the two?

File: node/subsystem-util/src/validator_discovery.rs
Comment (rphmeier) [inline node/subsystem-util/src/validator_discovery.rs:54]: I wonder if this function should have a `SessionIndex` attached to it.

File: node/subsystem-util/src/validator_discovery.rs
Comment (rphmeier) [inline node/subsystem-util/src/validator_discovery.rs:143]: missing docs on public item.

File: node/subsystem-util/src/validator_discovery.rs
Comment (rphmeier) [inline node/subsystem-util/src/validator_discovery.rs:112]: Should this be `#[must_use]` to avoid people dropping it without revoking?

File: node/subsystem/src/messages.rs
Comment (rphmeier) [inline node/subsystem/src/messages.rs:407]: I see, then let's ignore the session index elsewhere for now as is currently done.

**Review by rphmeier — APPROVED** [review]
General approach looks good. Are there associated guide changes?

File: node/subsystem-util/src/validator_discovery.rs
Comment (ordian) [reply node/subsystem-util/src/validator_discovery.rs:112]: Actually, dropping this struct will also result in revoking.
Previously in this PR, the behavior was that the request was never revoked in this case, but I don't think there is a use-case for that, is there?
Will update the docs.

File: primitives/src/v1.rs
Comment (ordian) [inline primitives/src/v1.rs:56]: any thoughts on this?

File: node/subsystem/src/messages.rs
Comment (ordian) [reply node/subsystem/src/messages.rs:407]: Yeah, what I'm worried about though is breaking changes. It's not clear how to make the API future-proof though.

File: node/network/bridge/src/validator_discovery.rs
Comment (montekki) [inline node/network/bridge/src/validator_discovery.rs:248]: TODOs need resolving?

**Review by montekki — APPROVED** [review]

File: node/network/bridge/src/validator_discovery.rs
Comment (ordian) [reply node/network/bridge/src/validator_discovery.rs:248]: see the last commit

**Comment by ordian** [comment]
Not sure if we want to merge it now or wait till 0.8.25 branches out.
Also does it need `burnin`?

**Comment by rphmeier** [comment]
I don't think this needs any burn-in. I'm not too worried about breaking changes before any point where we deploy this on Kusama or Polkadot.

**Comment by ordian** [comment]
To clarify my worry:
if we merge the PR now, it will get deployed on Kusama and Polkadot in 0.8.25 (if I understand correctly) 
this PR introduces a runtime API call `validator_discovery` that doesn't take `SessionIndex` into account, so it will be changed sometime in the future anyway
if we don't merge it now, there is a slight chance that we will fix the API before the next release.

**Comment by rphmeier** [comment]
> if we merge the PR now, it will get deployed on Kusama and Polkadot in 0.8.25 (if I understand correctly)
this PR introduces a runtime API call validator_discovery that doesn't take SessionIndex into account, so it will be changed sometime in the future anyway

There is a dummy implementation of the v1 runtime APIs including this new set of calls on Polkadot and Kusama. I figure that's OK as long as the node-side is resilient to runtime API errors, so it can author blocks even if a runtime API call fails. Runtime API call failure is really the only type of failure we have to worry about, as long as the dummy API is OK for authoring under the overseer. It would happen because we'd update the runtime API definition to be different than what the runtime actually exposes.

File: node/network/bridge/src/validator_discovery.rs
Comment (bkchr) [inline node/network/bridge/src/validator_discovery.rs:143]: Some docs what this is doing?

**Pull request merged**
